### DISCUSSION TITLE
Apache documentation

### DISCUSSION THREAD

**Discussion by czupet** [dotnet/yarp#421] [General] [answered]
Does anyone have an Apache documentation on Linux or an example?
My test app works on IIS, but not Linux.  I get Error: connect ECONNREFUSED 10.200.15.27:5000, but the same app with the same appsettings.json works on Windows and IIS.

**Comment by Tratcher** [comment]
Have you tried these? https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-apache?view=aspnetcore-3.1

**Comment by czupet** [comment]
Yes, but I will doublecheck, maybe I missed something - obviously.
Thanks for the fast response and the link.

**Comment by czupet** [comment]
I really don't know what I am doing when it comes to Apache and reverse proxy. I have this running on IIS, but can not get it to work on Apache on Linux. Getting the 404 error. 
I appreciate any help, I've been stuck on this for several days now.
I read the documentation at https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-apache?view=aspnetcore-3.1
but nothing seems to work. (Probably something really simple)

here's my /etc/httpd/conf.d/wsiproxy.conf

<VirtualHost *:80>
    ProxyPreserveHost On
    ProxyPass / http://127.0.0.1:5000/
    ProxyPassReverse / http://127.0.0.1:5000/
    ServerName www.example.com
    ServerAlias *.example.com
</VirtualHost>

here's my appsettings.json

{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "AllowedHosts": "*",
  "ReverseProxy": {
    "Routes": [
      {
        "RouteId": "app2",
        "ClusterId": "cluster1",
        "Match": {
          "Methods": [ "GET", "POST" ],
          "Path": "/wsirates/"
        }
      },

      {
        "RouteId": "route2",
        "ClusterId": "cluster2",
        "Match": {
          "Methods": [ "GET", "POST" ],
          "Path": "/wsirates3/"
        }
      }
    ],
    "Clusters": {
      "cluster1": {
        "LoadBalancing": {
          "Mode": "Random"
        },
        "Destinations": {
          "cluster1_destination1": {
            "Address": "http://reports04/Test/api/rates/"
          },
          "cluster1_destination2": {
            "Address": "http://reports04/Test/api/rates/"
          }
        }
      },
      "cluster2": {
        "Destinations": {
          "cluster2_destination1": {
            "Address": "http://reports04/Test/api/rates/"
          }
        }
      }
    }
  }
}

**Reply by samsp-msft** [reply]
What are you trying to accomplish? The way that YARP is designed, its intended to be hosted in the Kestrel web server. The doc you referenced for use with Apache is for earlier deployments of Kestrel where we were not yet confident that kestrel was fully ready to be internet facing, with .NET 3+ we are. To use YARP on linux, you should not be trying to put apache in front of it - that is adding another level of indirection and connections.

**Comment by czupet** [comment]
I am trying to use YARP on Linux.  
Currently, everything is on the internal network, but my intent is to have it facing the Internet.
1) an external app will call YARP.  http://www.mywebserver.com/wsirates
2) YARP will forward this call to my internal web server and the internal web server will return some data.

This is working on the intranet with YARP installed on IIS.
I am trying to get it to run on Linux and every documentation I have read it says you should have Apache or some other webserver running in front of Kestrel.

So how do I set this up? Is there any documentation to help me with that?

Thank You

**Reply by Tratcher** [reply]
> I am trying to get it to run on Linux and every documentation I have read it says you should have Apache or some other webserver running in front of Kestrel.

That guidance has been outdated since 2.0. Kestrel is now supported as a public edge server.

**Comment by czupet** [comment, marked as answer]
Thanks for your help.
I figured it out.  The problem was I had to tell Kestrel to accept requests from other IPs besides the localhost

public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseUrls("http://0.0.0.0:5000").UseStartup<Startup>();
                });

**Comment by samsp-msft** [comment]
@tratcher - should we change the docs for our sample to make this clearer that its only allowing connections from localhost? It seems that this is part of the ASP.NET project template, and may not be obvious to people as they move from local machine to deployment scenarios.

**Reply by Tratcher** [reply]
I think this would be covered under https://github.com/microsoft/reverse-proxy/issues/183, documenting recommended configurations for servers.

Yes Kestrel does default to `https://localhost:5001/` and `http://localhost:5000/` if no config is provided. AspNet logs at startup the endpoints its bound to.

Note the defaults rarely come into play, especially in the tooling scenarios. For example it's common to set the bindings in the launchsettings.json file used by VS or dotnet run.
https://github.com/dotnet/aspnetcore/blob/4dfa2c719f85b2b10fea17a9e9cfc0a19cbc4d46/src/Security/Authentication/samples/SocialSample/Properties/launchSettings.json#L24

Some of our samples already call out the kestrel config like this one:
https://github.com/microsoft/reverse-proxy/blob/b6353e868ba70db9bab4a97d9972b807aaaa145b/samples/ReverseProxy.Config.Sample/appsettings.json#L10-L19

**Comment by czupet** [comment]
That would help people like me because I had no idea about this so I couldn't even search for this because I did not know what to search for.  For seasoned people would be just an obvious statement.

Thank You.
